### ISSUE TITLE
More live points make result worse

### ISSUE THREAD

**Issue by hippke** [joshspeagle/dynesty#376]
Thank you for making this great package! Perhaps my issue is not a bug but a misunderstanding of the dynesty parameters.

My problem (exomoon search) has ~15 dimensions. To systematically find the best sampling method, I use synthetic data - I know the ground truth. With UltraNest, the RegionSliceSampler/Stepsampler works well with a few thousand live points (all else defaults). The problem here is that it doesn't parallelize very well. That's a strength of dynesty.

With dynesty, I find that the dynamic method with slices and multiple ellipsoids works best. However, I struggle to find the best combination of the number of slices and live points.

I tested 5, 10, 15 and 20 slices. For each, I tested a series of live points (hundreds to thousands). I would expect that for more slices and more live points, the results would get better. After all, it should lead to a more thorough exploration. However, that seems not to be the case.

As an example, I made a figure for the case where I keep `slices=15` and explore different values for `nlive`. I repeat each combination of parameters 10 times. Afterwards, I plot the number of model evaluations (essentially my runtime) as a function of the logz, which I consider a good measure of the success of a run. A perfect sampling run should yield `logz~170`. 

```
sampler = DynamicNestedSampler(
            log_likelihood,
            prior_transform,
            ndim=ndim,
            bound='multi',
            sample='slice',
            slices=15,
            nlive=3000
            )
```
Surprisingly, the result does not get better with more live points. Why is that?
![1](https://user-images.githubusercontent.com/11958117/170497530-5065e620-93ac-44cc-9522-501eb70b119f.png)

**Comment by segasai** [comment]
I am assuming that you are using the very latest dynesty version ? 1.2.2 ? ( if not please update and repeat). 

Here're my thoughts. 

* If you are 100% sure that the correct answer is -170, the fact that the result doesn't converge when increasing nslice and nlive  would be a bug of dynesty. 
* In general my expectation of the dependence on the nslice is the following. If nslice is too small, the new samples at each iteration are not independent and therefore the shrinking of posterior volume will be incorrect and therefore the result can be biased. But at some point it should saturate. (also I don't quite have a good feeling whether large nlive can rectify the small nslice problem)
* The nlive in my mind affects several aspects of the fit. If nlive is small you can easily miss some modes of the distribution (if it's multimodal) and also can introduce stochasticity as some modes may be lost during sampling, leading to inaccurate weights between modes if they don't have enough live-points. Also if nlive is small and rslice is small, I can imagine possible difficulties in complex areas of the posterior (i.e when it' significantly more curved/narrower than other parts)

What's happening here is a bit hard to tell.  I think you may want to bump the number of slices higher and see what's happening. If I had to guess -- that's the answer, i.e. nslice=15 is still not enough (in fact I think the default adopted by dynesty is 5+ndim or something like that)

**Comment by segasai** [comment]
> From the [documentation](https://dynesty.readthedocs.io/en/latest/faq.html#sampling-questions): "In general we find that the default (and conservative) `slices=5` is robust under a wide variety of circumstances."

Just on this point -- this is talking about the 'slice' sampler, where the the actual number of slices is slices*ndim. The rslice sampler doesn't multiply by ndim. 
(the docs need to be definitely clarified on this)

Regarding the rest, I need to think about this, I don't have an immediate idea there.
Could you try the same run  with the single ellipsoidal bound ? I know it'll be significantly less efficient but that would separate potential problems related to multiple ellipsoids..
Also, you could you try rwalk as well ? I have in general more trust in rslice, but trying a different sampler could help with the diagnosis.

**Comment by segasai** [comment]
Also, if you can share (even privately) the code that can be run that may be helpful to debug the issue.

**Comment by hippke** [comment]
As suggested, I have explored the following options:
- Single ellipsoidal bound with [15, 20, 25] slices and [150, 300, 600, 1200, 2400] nlive
- Multi ellipsoidal bound with large [50, 75, 100] slices and [150, 300, 600, 1200, 2400] nlive
- Multi ellipsoidal bound rwalk with slices [5, 10, 15, 20, 25, 50, 100, 150] and [150, 300, 600, 1200, 2400] nlive

Not all has completed yet, so no figures yet, but it is already clear that these options are no improvements.

What is even more confusing: I have repeated yesterday's "best" run (15 slices, 375 nlive) and it is much worse! I have not changed the script at all. No idea why it fluctuates so much.

A minimal example is on my [Github gist](https://gist.github.com/hippke/b555ec14bbcc406f80b58fbc78437bee). You need `pip install pandoramoon` which is my [Pandora](https://github.com/hippke/Pandora) package to create Planet+Moon light curves (more in [this paper](https://arxiv.org/abs/2205.09410)). 

Thank you so much for your help!

**Comment by segasai** [comment]
Thanks for the example, that's very helpful, I can run it. 
And where does the correct answer of -170 comes from ? I.e. I understand that this is mock data where you know the true value of parameters, but it's unclear how you know exactly computed the "true" logz value. Or is it just the value that ultranest gives you ?

**Comment by hippke** [comment]
Yes, with a very deep UltraNest sampling I get logz=~-170. 
The posteriors are then within 1-sigma of the true values. I typically see (in this example) reasonable cornerplots for logz values in the range -170 to -180.
In dynesty runs with typical logz~ -200, the posteriors are just wrong: Way off the true values, and with wrong uncertainties.

**Comment by segasai** [comment]
okay, It seems to me, that the likely underlying problem is just missing a right narrow posterior mode completely rather than getting a logz wrong. Can you show your "correct" posterior  ? Or 1-99 percentiles in each parameter ? I.e. I'm interested if 
the posteriors dynesty gets have no overlap with "correct" posteriors or not (that would support missing the mode hypothesis).  
(the reason why I'm saying "correct" in quotation marks, is because while even if all the code doing the computations is accurate, it's still possible to have a posterior peaking at "wrong" values, if the likelihood is not informative enough and you are prior driven. And I obviously don't know your problem well enough to see if that's the case here.) 
Usually a good test of that theory would be to increase the infomativeness of your data, i.e. increase the number of datapoints and decrease the error-bars.

**Comment by hippke** [comment]
Yes, it seems that the correct narrow posterior mode is missed. 
Attached is a cornerplot with "correct" posteriors. I didn't save the numbers but can re-run if needed.
[cornerplot_injection_retrieval.pdf](https://github.com/joshspeagle/dynesty/files/8791628/cornerplot_injection_retrieval.pdf).

To increase the informativeness of my data, you can
1. change `noise_level = 500e-6` (line 58) to a smaller value. For example, setting `noise_level = 100e-6` increases SNR by 5x.
2. In addition, the temporal resolution can be increased (more data points per unit time). In line 42, you can change `params.cadences_per_day = 48` to a higher value. This comes at linear increase of compute.

**Comment by hippke** [comment]
One more experiment where I vary `nlive` and keep all else constant for `sample="slice"` and all else defaults.
```
sampler = NestedSampler(
    log_likelihood,
    prior_transform,
    ndim=15,
    sample='slice',
    nlive=nlive
    )
```
As expected, the number of model evaluations increases with more live points:
![1](https://user-images.githubusercontent.com/11958117/170875521-b08d4c8d-d868-4591-9c96-1b730cd434b9.png)

The logz however does not improve:
![2](https://user-images.githubusercontent.com/11958117/170875564-95b9aed8-6a02-4909-9f6b-6c093629730a.png)

**Comment by hippke** [comment]
What seems to work better: The combination of dynamic sampling, `bound="balls"` and `bootstrap=20`.
Then I get reasonable results with `nlive>~2000'

```
sampler = DynamicNestedSampler(
    log_likelihood,
    prior_transform,
    ndim=15,
    bound='balls',
    sample='slice',
    nlive=2000,
    bootstrap=20
    )
```

**Comment by segasai** [comment]
I still don't have a full understanding of what's going on but
* I don't think looking at logz value is a useful diagnostic of anything here, but I agree there are issues with the posterior. 
* It seems to me that the discreteness of the posterior in per_moon is likely a problem -- you can see it in your pdf -- basically the posterior consists of a large number very narrow strips. That's an extremely difficult thing to sample with metropolis-hastings like moves and certainly more suited for rejection type sampling (uniform/balls/boxes samplers)
* Do you know what's the highest logl values in your 'correct' posterior ? From what I understand we are not missing a particularly high logl value, it's really the issue is with those narrow strips.

Unfortunately the uniform sampler with ellipsoidal bounds seems quite inefficient for this problem (which is probably a separate issue, because it's not parallellized, and in this case the ellipsoid approximation of the posterior is largely outside the cube and so a lot of samples are wasted this way)

**Comment by hippke** [comment]
Yes, the posterior of `per_moon` is typically strip(s) - one or many. That's due to the physics of the planet-moon system, where certain combinations of planetary and moon periods create a "strobe" effect of signals. On the other hand, the uncertainty with realistic SNR (and few transits) is often quite large. In our example, the physical limit is [1..30] days, and the posterior is 5 +- 0.3 days. With infinite SNR, the posterior would be a finite number of infinitely thin strips.

I tried `unif` sampling but it gets stuck. Slice sampling with cubes work better than ellipsoids, as you describe it. Perhaps I could live with that. On the other hand, I really would want the result to get better with more live points (and it doesn't).

Is there a different package that I should try out for rejection sampling? Or can it be tested with dynesty?

How do I calculate logl for the correct posterior? Sorry, I'm new to this adventure :-)

**Comment by segasai** [comment]
Okay.
(after some time spending on this code)

First your test is somewhat problematic -- you test on different data, which prevents any comparisons from run to run. 

Here's a modified script where data is generated for one given data seed and there is a separate fitting seed 

```
import pandoramoon as pandora
from pandoramoon.helpers import ld_convert
import numpy as np
from dynesty import DynamicNestedSampler, NestedSampler
from multiprocessing import Pool, cpu_count
import dynesty.utils as dyutil
# Injection part
# ==============

R_sun = 696_342_000


def get_data(data_seed=1):
    # Call Pandora and get model with these parameters
    params = pandora.model_params()
    params.R_star = R_sun  # [m]
    params.u1 = 0.4089
    params.u2 = 0.2556

    # Planet parameters
    params.per_bary = 365.25  # [days]
    params.a_bary = 215  # [R_star]
    params.r_planet = 0.1  # [R_star]
    params.b_bary = 0.3  # [0..1]
    params.t0_bary = 11  # [days]
    params.t0_bary_offset = 0.01  # [days]
    params.M_planet = 1.8986e+27  # [kg]
    params.w_bary = 0  # [deg]
    params.ecc_bary = 0.0  # [0..1]

    # Moon parameters
    params.r_moon = 0.03526  # [R_star]
    params.per_moon = 5  # [days]
    params.tau_moon = 0.07  # [0..1]
    params.Omega_moon = 0  # [0..180]
    params.i_moon = 90  # [0..180]
    params.e_moon = 0  # [0..1]
    params.w_moon = 0  # [deg]
    params.M_moon = 0.05395 * params.M_planet  # [0..1]

    # Other model parameters
    params.epochs = 3  # [int]
    params.epoch_duration = 2  # [days]
    params.cadences_per_day = 48  # [int]
    params.epoch_distance = 365.26  # [days]
    params.supersampling_factor = 1  # [int]
    params.occult_small_threshold = 0.1  # [0..1]
    params.hill_sphere_threshold = 1.2

    # Obtain time grid
    time = pandora.time(params).grid()

    # Define model
    model = pandora.moon_model(params)

    # Evaluate model for each point in time grid
    injected_flux_total, injected_flux_planet, injected_flux_moon = model.light_curve(
        time)

    # Create noise and merge with flux
    noise_level = 500e-6  # Gaussian noise to be added to the generated data
    rng = np.random.default_rng(data_seed)
    noise = rng.normal(0, noise_level, len(time))
    testdata = noise + injected_flux_total
    yerr = np.full(len(testdata), noise_level)
    return params, testdata, time, yerr


"""
# Plot synthetic data with and without noise
import matplotlib.pyplot as plt
plt.plot(time, injected_flux_planet, color="blue")
plt.plot(time, injected_flux_moon, color="red")
plt.plot(time, injected_flux_total, color="black")
plt.scatter(time, testdata, color="black", s=0.5)
plt.xlabel("Time (days)")
plt.ylabel("Relative flux")
plt.show()
"""

# Retrieval part
# ==============


def prior_transform(cube):
    # The argument: "unit-hypercube" consists of values [0,1]
    # We have to convert them to physical scales
    p = cube.copy()
    p[0] = cube[0] * 0.8 * R_sun + 0.4 * R_sun  #  R_star [m]
    p[1] = cube[1] * 10 + 360.25  # per_bary [days]
    p[2] = cube[2] * 10 + 210  # a_bary [R_star]
    p[3] = cube[3] * 0.04 + 0.07  # r_planet [R_star]
    p[4] = cube[4] * 1  # b_bary [0..1]
    p[5] = cube[5] * 0.1 - 0.05  # t0_bary_offset [days]
    p[6] = cube[6] * 1e27 + 1.5e27  # M_planet [kg]
    p[7] = cube[7] * 0.06 + 0.01  # r_moon  [R_star]
    p[8] = cube[8] * 50  # per_moon  [days]
    p[9] = cube[9] * 1  # tau [0..1]
    p[10] = cube[10] * 180  # Omega_moon [0..180]
    p[11] = cube[11] * 180  # i_moon [0..180]
    p[12] = cube[12] * 1e27 + 1  # M_moon [kg]
    p[13] = cube[13]  # LD q1 [0..1]
    p[14] = cube[14]  # LD q2 [0..1]
    return p


def log_likelihood(p, data):
    params, testdata, time, yerr = data
    # Convert q priors to u LDs (Kipping 2013)
    q1 = p[13]
    q2 = p[14]
    u1, u2 = ld_convert(q1, q2)

    # Calculate pandora model with trial parameters
    _, _, flux_trial_total, _, _, _, _ = pandora.pandora(
        R_star=p[0],
        u1=u1,
        u2=u2,

        # Planet parameters
        per_bary=p[1],
        a_bary=p[2],
        r_planet=p[3],
        b_bary=p[4],
        w_bary=0,
        ecc_bary=0,
        t0_bary=params.t0_bary,
        t0_bary_offset=p[5],
        M_planet=p[6],

        # Moon parameters
        r_moon=p[7],
        per_moon=p[8],
        tau_moon=p[9],
        Omega_moon=p[10],
        i_moon=p[11],
        ecc_moon=0,
        w_moon=0,
        M_moon=p[12],

        # Other model parameters
        epoch_distance=params.epoch_distance,
        supersampling_factor=1,
        occult_small_threshold=0.01,
        hill_sphere_threshold=1.0,
        numerical_grid=25,
        time=time,
        # cache=cache  # Can't use cache because free LDs
    )
    loglike = -0.5 * np.nansum(((flux_trial_total - testdata) / yerr)**2)

    return loglike


parameters = [
    'R_star', 'per_bary', 'a_bary', 'r_planet', 'b_bary', 't0_bary_offset',
    'M_planet', 'r_moon', 'per_moon', 'tau', 'Omega_moon', 'i_moon', 'M_moon',
    'q1', 'q2'
]

wrapped_params = [
    False,
    False,
    False,
    False,
    False,
    False,
    False,
    False,
    False,
    True,  # tau
    False,
    False,
    False,
    False,
    False
]


def doit(seed_data=1,
         seed_fit=0,
         slices=20,
         nlive=500,
         sample='rslice',
         bound='multi'):
    (params, testdata, time, yerr) = get_data(seed_data)
    rstate = np.random.default_rng(seed_fit)
    with Pool(cpu_count()) as executor:
        sampler = DynamicNestedSampler(
            log_likelihood,
            prior_transform,
            ndim=15,
            sample=sample,
            # bound='multi',
            # sample='rslice',
            # slices=slices,
            nlive=nlive,
            pool=executor,
            queue_size=cpu_count(),
            # bootstrap=0
            rstate=rstate,
            logl_args=([params, testdata, time, yerr], ),
        )
        sampler.run_nested()
        res = sampler.results
    chains = dyutil.resample_equal(res['samples'],
                                   np.exp(res['logwt'] - res['logz'][-1]))
    return chains
```
This actually allows better understanding of what's happening

Here's one posterior
![qq](https://user-images.githubusercontent.com/109513/171052300-cbd50e24-2f1d-456a-a199-fc09062f464d.png)

which looks pretty genuine. The only thing that happens here is here only one of those narrow stripes is picked up. 
So that's really the underlying problem -- those stripes. Since their width is < 1e-3 and there are many of those they are easily missed.

Clearly uniform sampling would be appropriate here, but you have a problematic parameterization. you posterior is very thin in some dimensions and takes the full box in some other dimensions, so what would help here 
* removing sharp edges in the prior so the posterior is closer to an actual ellipsoid will make uniform sampler work better. You can replace some priors from u() to N()
* Use ellipsoidal bounding/clustering only for some dimensions, while uniformly sampling others (it can be done with ncdim parameter) 

I tried to swap per_bary and a couple of other parameters to make them first arguments and use small ncdim with the uniform sampler -- that clearly makes the fit progress till quite high log-likelihood (but i don't have enough patience to wait till the full run)

**Comment by segasai** [comment]
> * In the cornerplot, it shows a narrow posterior for a wrong `per_moon` (true: 5, found: 5.214 +- 0.01). It seems like the sampler got stuck in one of the less significant strips?

It is not clear which strip is more significant, but it caught only one in this case.

> * I'd love to try sampling uniformely for `per_moon`, and with other methods for the other parameters. Can you share a code snippet to get started? I've never worked with  `ncdim` before. I'll expend considerable compute to explore this method.

Here is the ncdim variant (I basically swapped some parameters around).
The whole point is have first parameters where ellipsoidal parametrization makes sense and should help increase sampling efificiency and the other dimensions basically sample uniformly  [0,1].
```
import pandoramoon as pandora
from pandoramoon.helpers import ld_convert
import numpy as np
from dynesty import DynamicNestedSampler, NestedSampler
from multiprocessing import Pool, cpu_count
import dynesty.utils as dyutil
# Injection part
# ==============

R_sun = 696_342_000


def get_data(data_seed=1):
    # Call Pandora and get model with these parameters
    params = pandora.model_params()
    params.R_star = R_sun  # [m]
    params.u1 = 0.4089
    params.u2 = 0.2556

    # Planet parameters
    params.per_bary = 365.25  # [days]
    params.a_bary = 215  # [R_star]
    params.r_planet = 0.1  # [R_star]
    params.b_bary = 0.3  # [0..1]
    params.t0_bary = 11  # [days]
    params.t0_bary_offset = 0.01  # [days]
    params.M_planet = 1.8986e+27  # [kg]
    params.w_bary = 0  # [deg]
    params.ecc_bary = 0.0  # [0..1]

    # Moon parameters
    params.r_moon = 0.03526  # [R_star]
    params.per_moon = 5  # [days]
    params.tau_moon = 0.07  # [0..1]
    params.Omega_moon = 0  # [0..180]
    params.i_moon = 90  # [0..180]
    params.e_moon = 0  # [0..1]
    params.w_moon = 0  # [deg]
    params.M_moon = 0.05395 * params.M_planet  # [0..1]

    # Other model parameters
    params.epochs = 3  # [int]
    params.epoch_duration = 2  # [days]
    params.cadences_per_day = 48  # [int]
    params.epoch_distance = 365.26  # [days]
    params.supersampling_factor = 1  # [int]
    params.occult_small_threshold = 0.1  # [0..1]
    params.hill_sphere_threshold = 1.2

    # Obtain time grid
    time = pandora.time(params).grid()

    # Define model
    model = pandora.moon_model(params)

    # Evaluate model for each point in time grid
    injected_flux_total, injected_flux_planet, injected_flux_moon = model.light_curve(
        time)

    # Create noise and merge with flux
    noise_level = 500e-6  # Gaussian noise to be added to the generated data
    rng = np.random.default_rng(data_seed)
    noise = rng.normal(0, noise_level, len(time))
    testdata = noise + injected_flux_total
    yerr = np.full(len(testdata), noise_level)
    return params, testdata, time, yerr


"""
# Plot synthetic data with and without noise
import matplotlib.pyplot as plt
plt.plot(time, injected_flux_planet, color="blue")
plt.plot(time, injected_flux_moon, color="red")
plt.plot(time, injected_flux_total, color="black")
plt.scatter(time, testdata, color="black", s=0.5)
plt.xlabel("Time (days)")
plt.ylabel("Relative flux")
plt.show()
"""

# Retrieval part
# ==============


def prior_transform(cube):
    # The argument: "unit-hypercube" consists of values [0,1]
    # We have to convert them to physical scales
    p = cube.copy()
    p[0] = cube[0] * 10 + 360.25  # per_bary [days]
    p[1] = cube[1] * 180  # i_moon [0..180]
    p[2] = cube[2] * 10 + 210  # a_bary [R_star]
    p[3] = cube[3] * 0.04 + 0.07  # r_planet [R_star]
    p[4] = cube[4] * 1  # b_bary [0..1]
    p[5] = cube[5] * 0.1 - 0.05  # t0_bary_offset [days]
    p[6] = cube[6] * 1e27 + 1.5e27  # M_planet [kg]
    p[7] = cube[7] * 0.06 + 0.01  # r_moon  [R_star]
    p[8] = cube[8] * 50  # per_moon  [days]
    p[9] = cube[9] * 1  # tau [0..1]
    p[10] = cube[10] * 180  # Omega_moon [0..180]
    p[11] = cube[11] * 0.8 * R_sun + 0.4 * R_sun  #  R_star [m]
    p[12] = cube[12] * 1e27 + 1  # M_moon [kg]
    p[13] = cube[13]  # LD q1 [0..1]
    p[14] = cube[14]  # LD q2 [0..1]
    return p


def log_likelihood(p, data):
    params, testdata, time, yerr = data
    # Convert q priors to u LDs (Kipping 2013)
    q1 = p[13]
    q2 = p[14]
    u1, u2 = ld_convert(q1, q2)

    # Calculate pandora model with trial parameters
    _, _, flux_trial_total, _, _, _, _ = pandora.pandora(
        R_star=p[11],
        u1=u1,
        u2=u2,

        # Planet parameters
        per_bary=p[0],
        a_bary=p[2],
        r_planet=p[3],
        b_bary=p[4],
        w_bary=0,
        ecc_bary=0,
        t0_bary=params.t0_bary,
        t0_bary_offset=p[5],
        M_planet=p[6],

        # Moon parameters
        r_moon=p[7],
        per_moon=p[8],
        tau_moon=p[9],
        Omega_moon=p[10],
        i_moon=p[1],
        ecc_moon=0,
        w_moon=0,
        M_moon=p[12],

        # Other model parameters
        epoch_distance=params.epoch_distance,
        supersampling_factor=1,
        occult_small_threshold=0.01,
        hill_sphere_threshold=1.0,
        numerical_grid=25,
        time=time,
        # cache=cache  # Can't use cache because free LDs
    )
    loglike = -0.5 * np.nansum(((flux_trial_total - testdata) / yerr)**2)

    return loglike


parameters = [
    'R_star', 'per_bary', 'a_bary', 'r_planet', 'b_bary', 't0_bary_offset',
    'M_planet', 'r_moon', 'per_moon', 'tau', 'Omega_moon', 'i_moon', 'M_moon',
    'q1', 'q2'
]

wrapped_params = [
    False,
    False,
    False,
    False,
    False,
    False,
    False,
    False,
    False,
    True,  # tau
    False,
    False,
    False,
    False,
    False
]


def doit(seed_data=1,
         seed_fit=0,
         slices=20,
         nlive=500,
         sample='rslice',
         bound='multi'):
    (params, testdata, time, yerr) = get_data(seed_data)
    rstate = np.random.default_rng(seed_fit)
    with Pool(cpu_count()) as executor:
        sampler = DynamicNestedSampler(
            log_likelihood,
            prior_transform,
            ndim=15,
            ncdim=2,
            sample=sample,
            # bound='multi',
            # sample='rslice',
            # slices=slices,
            nlive=nlive,
            pool=executor,
            queue_size=cpu_count(),
            # bootstrap=0
            rstate=rstate,
            logl_args=([params, testdata, time, yerr], ),
        )
        sampler.run_nested()
        res = sampler.results
    chains = dyutil.resample_equal(res['samples'],
                                   np.exp(res['logwt'] - res['logz'][-1]))
    return chains

```

And here's the alternative approach which I think works the best -- basically here I've just ran the dynamic sampler and added 100 batches spanning the full posterior)

![qq2](https://user-images.githubusercontent.com/109513/171158810-842a96e3-f0d4-4a3b-9efe-cc6b099c0bc3.png)
This is basically equivalent to doing the sampling 100 times and then merging the results


Code: 
```
import pandoramoon as pandora
from pandoramoon.helpers import ld_convert
import numpy as np
from dynesty import DynamicNestedSampler, NestedSampler
from multiprocessing import Pool, cpu_count
import dynesty.utils as dyutil
# Injection part
# ==============

R_sun = 696_342_000


def get_data(data_seed=1):
    # Call Pandora and get model with these parameters
    params = pandora.model_params()
    params.R_star = R_sun  # [m]
    params.u1 = 0.4089
    params.u2 = 0.2556

    # Planet parameters
    params.per_bary = 365.25  # [days]
    params.a_bary = 215  # [R_star]
    params.r_planet = 0.1  # [R_star]
    params.b_bary = 0.3  # [0..1]
    params.t0_bary = 11  # [days]
    params.t0_bary_offset = 0.01  # [days]
    params.M_planet = 1.8986e+27  # [kg]
    params.w_bary = 0  # [deg]
    params.ecc_bary = 0.0  # [0..1]

    # Moon parameters
    params.r_moon = 0.03526  # [R_star]
    params.per_moon = 5  # [days]
    params.tau_moon = 0.07  # [0..1]
    params.Omega_moon = 0  # [0..180]
    params.i_moon = 90  # [0..180]
    params.e_moon = 0  # [0..1]
    params.w_moon = 0  # [deg]
    params.M_moon = 0.05395 * params.M_planet  # [0..1]

    # Other model parameters
    params.epochs = 3  # [int]
    params.epoch_duration = 2  # [days]
    params.cadences_per_day = 48  # [int]
    params.epoch_distance = 365.26  # [days]
    params.supersampling_factor = 1  # [int]
    params.occult_small_threshold = 0.1  # [0..1]
    params.hill_sphere_threshold = 1.2

    # Obtain time grid
    time = pandora.time(params).grid()

    # Define model
    model = pandora.moon_model(params)

    # Evaluate model for each point in time grid
    injected_flux_total, injected_flux_planet, injected_flux_moon = model.light_curve(
        time)

    # Create noise and merge with flux
    noise_level = 500e-6  # Gaussian noise to be added to the generated data
    rng = np.random.default_rng(data_seed)
    noise = rng.normal(0, noise_level, len(time))
    testdata = noise + injected_flux_total
    yerr = np.full(len(testdata), noise_level)
    return params, testdata, time, yerr


"""
# Plot synthetic data with and without noise
import matplotlib.pyplot as plt
plt.plot(time, injected_flux_planet, color="blue")
plt.plot(time, injected_flux_moon, color="red")
plt.plot(time, injected_flux_total, color="black")
plt.scatter(time, testdata, color="black", s=0.5)
plt.xlabel("Time (days)")
plt.ylabel("Relative flux")
plt.show()
"""

# Retrieval part
# ==============


def prior_transform(cube):
    # The argument: "unit-hypercube" consists of values [0,1]
    # We have to convert them to physical scales
    p = cube.copy()
    p[0] = cube[0] * 10 + 360.25  # per_bary [days]
    p[1] = cube[1] * 180  # i_moon [0..180]
    p[2] = cube[2] * 10 + 210  # a_bary [R_star]
    p[3] = cube[3] * 0.04 + 0.07  # r_planet [R_star]
    p[4] = cube[4] * 1  # b_bary [0..1]
    p[5] = cube[5] * 0.1 - 0.05  # t0_bary_offset [days]
    p[6] = cube[6] * 1e27 + 1.5e27  # M_planet [kg]
    p[7] = cube[7] * 0.06 + 0.01  # r_moon  [R_star]
    p[8] = cube[8] * 50  # per_moon  [days]
    p[9] = cube[9] * 1  # tau [0..1]
    p[10] = cube[10] * 180  # Omega_moon [0..180]
    p[11] = cube[11] * 0.8 * R_sun + 0.4 * R_sun  #  R_star [m]
    p[12] = cube[12] * 1e27 + 1  # M_moon [kg]
    p[13] = cube[13]  # LD q1 [0..1]
    p[14] = cube[14]  # LD q2 [0..1]
    return p


def log_likelihood(p, data):
    params, testdata, time, yerr = data
    # Convert q priors to u LDs (Kipping 2013)
    q1 = p[13]
    q2 = p[14]
    u1, u2 = ld_convert(q1, q2)

    # Calculate pandora model with trial parameters
    _, _, flux_trial_total, _, _, _, _ = pandora.pandora(
        R_star=p[11],
        u1=u1,
        u2=u2,

        # Planet parameters
        per_bary=p[0],
        a_bary=p[2],
        r_planet=p[3],
        b_bary=p[4],
        w_bary=0,
        ecc_bary=0,
        t0_bary=params.t0_bary,
        t0_bary_offset=p[5],
        M_planet=p[6],

        # Moon parameters
        r_moon=p[7],
        per_moon=p[8],
        tau_moon=p[9],
        Omega_moon=p[10],
        i_moon=p[1],
        ecc_moon=0,
        w_moon=0,
        M_moon=p[12],

        # Other model parameters
        epoch_distance=params.epoch_distance,
        supersampling_factor=1,
        occult_small_threshold=0.01,
        hill_sphere_threshold=1.0,
        numerical_grid=25,
        time=time,
        # cache=cache  # Can't use cache because free LDs
    )
    loglike = -0.5 * np.nansum(((flux_trial_total - testdata) / yerr)**2)

    return loglike


parameters = [
    'R_star', 'per_bary', 'a_bary', 'r_planet', 'b_bary', 't0_bary_offset',
    'M_planet', 'r_moon', 'per_moon', 'tau', 'Omega_moon', 'i_moon', 'M_moon',
    'q1', 'q2'
]

wrapped_params = [
    False,
    False,
    False,
    False,
    False,
    False,
    False,
    False,
    False,
    True,  # tau
    False,
    False,
    False,
    False,
    False
]


def doit(seed_data=1,
         seed_fit=0,
         slices=20,
         nlive=100,
         sample='rslice',
         bound='multi'):
    (params, testdata, time, yerr) = get_data(seed_data)
    rstate = np.random.default_rng(seed_fit)
    ##with Pool(cpu_count()) as executor:
    if True:
        sampler = DynamicNestedSampler(
            log_likelihood,
            prior_transform,
            ndim=15,
            sample=sample,
            # bound='multi',
            # sample='rslice',
            # slices=slices,
            nlive=nlive,
            # pool=executor,
            #queue_size=cpu_count(),
            # bootstrap=0
            rstate=rstate,
            logl_args=([params, testdata, time, yerr], ),
        )
        sampler.run_nested(maxbatch=0)
        for i in range(100):
            sampler.add_batch(mode='full', nlive=nlive)
        res = sampler.results
    chains = dyutil.resample_equal(res['samples'],
                                   np.exp(res['logwt'] - res['logz'][-1]))
    return chains
```

Nonetheless you should think about reparametrising your model. For example you want to avoid the wrapping in angle. as that will create unnecessary multiple modes. Again, sharp prior edges are not good either.

**Comment by hippke** [comment]
Hey, that's good progress! Some quick questions:

- The first script doesn't work: "ncdim unsupported for slice sampling". What do I need to change?
- Is there a definition of `ncdim`? I couldn't find it in the documentation. Does `ncdim=2` mean it's uniform for the first two parameters and ellipsoid for the others?
- For the parameters with wrapping: Would it be acceptable to declare them via `reflective=[1,2,...]` or is that still bad?

**Comment by hippke** [comment]
With ~100 batches I get excellent coverage of the entire landscape. Very good approach!
To achieve good parallelization, I tried 100x `rlist = pool.map(run_one_batch, params)`, and merge afterwards with `dyfunc.merge_runs(rlist)`. That works, but...
- Is there a more elegant way to run 100 batches in parallel? 
- One issue with the merging is that it returns a list of values. So I get 100 logz values. Is there a way to condense them to a single logz?

**Comment by segasai** [comment]
* If you want to run 100 batches in parallel, completely independently, you should just run static runs and merge them 
*  I think merge_run should return one single logz 
* Also, when merging runs that are probing different modes, I haven't seen any theoretical proof of correctness of this , so I would be reluctant to put a lot of of weight on the logz in this case (IMO).

**Comment by hippke** [comment]
The individual batches work great. Is there a way to pass an additional parameter through the sampler into either the `log_likelihood` or `prior_transform` functions? With that, I could partition the moon period for each batch (i.e., reduce the prior). For example, I could use 50 batches and each scans 1/50 of the prior parameter space (while leaving all other parameters with their full priors).

Something like this:
```
batch_id = 10
sampler = NestedSampler(
    log_likelihood,
    prior_transform(batch_id),
    )
```

**Comment by segasai** [comment]
You can always make a logl or any other function a callable class with parameters.
 Regarding decomposing your prior into sections, you have to be careful there in weighting samples from those by evidence of that section.

**Comment by segasai** [comment]
I'll close this for now. Feel free to open/reopen if there is still a specific issue that needs resolution.
